### ISSUE TITLE
Bug: Custom operator with manual InstallPlanApproval fails to restore to healthy state.

### ISSUE THREAD

**Issue by lachm99** [openshift/oadp-operator#623]
### Contact Details

lachlan.macdonald@ibm.com

### Describe bug

Restoring a namespace which contains a custom operator with manual InstallPlanApproval fails to correctly link the Subscription, CSV, and InstallPlans.  
Instead, although the operator resources and pods are restored correctly, the Operator itself does not restore to a healthy state.  
Under certain circumstances, restore does succeed - but never by correctly linking all the dependent resources, but rather because the Subscription appears to generate new installPlans (the restored installPlans remain as broken relics) and a new CSV.

> **Note**: Restic, Velero, and OADP all report the restore as successful, because all resources are theoretically restored, the problem is that the 'restored' operator is not healthy - all the component resources don't link.

**Reproduce:**
- Have a custom CatalogSource in the openshift-marketplace project.
- Generate a stock OperatorGroup in the target project.
- Apply a subscription for your custom Operator in the target project
    - **Manual** installPlanApproval
    - Source must be the custom CatalogSource
- Approve and complete operator installation.
- Once the operator is in a healthy state, backup the project with the following settings using OADP.
    ```yaml
    apiVersion: velero.io/v1
    kind: Backup
    metadata:
      namespace: openshift-adp
      name: <name>
    spec:
      defaultVolumesToRestic: true
      storageLocation: oadp-dpa-1
      includedNamespaces:
        - <target-project>
     ```
- Once the backup is completed, completely clear the resources and delete the project.
- Restore from backup

### What happened?

I was testing backup/restore of a proprietary application, which is managed by a custom operator.
I conducted a backup of the project (including restic) while everything was in a healthy state.
Restoring this backup caused all the pods to reboot correctly, and for the application itself to work.
**However**, the operator did not restore properly, so nothing was in control of its restored dependent resources.
The restored CSV showed up in the InstalledOperators, but did not contain the subscription. The subscription resource **was** restored, but wasn't inherited by the CSV. It also displayed _unknown error_ under the upgrade status, and had no InstallPlan associated.  
Additionally, the CSV status claims `csv in namespace with no operatorgroups`, even though the operatorgroup of interest was successfully restored.

By doing the restore in stages (restoring namespace, operatorgroup, install plan first, then restoring only the subscription, then everything else), it appears I can brute force the subscription to generate new InstallPlans and a new CSV for itself, but the old installplans remain broken and unlinked. This also only works under certain circumstances, and when the operator version had an update pending or installed before the backup took place, I could not find any ordering that worked.

### OADP Version

1.0.1 (Stable)

### OpenShift Version

4.8

### New issue

- [X] This issue is new

**Comment by kaovilai** [comment]
OADP cannot currently restore operator workloads without much tinkering.

You can try to restore just a few items in the order that you installed your custom operator workload.
In your own words
1. restore "... stock OperatorGroup in the target project."
2. restore "subscription for your custom Operator in the target project"
  - Manual installPlanApproval
Anything after operator installation such as custom resources are not predictable as each operator has its own behavior when custom resources are created and the order it expects items to be created.

That's why plugins such as https://github.com/kubevirt/kubevirt-velero-plugin exists.

If you expect to backup and restore this operator on a regular, consider making a velero plugin to handle this operator.

**Comment by kaovilai** [comment]
TL;DR velero restore, recreates all items in the backup. It is not responsible for making sure the status fields are healthy on all types of workload. Beyond operators, other custom resource controllers may have custom admission behaviors that velero doesn't know.

Similar issue:
https://github.com/vmware-tanzu/velero/issues/4847
